rust/dyn-ddlog/src/ast/tree.rs

```diff
@@ -0,0 +1,651 @@
+    }
+
+    pub fn greater_equal(lhs: Expr<I>, rhs: Expr<I>) -> Self {
+        Self::cmp(lhs, CmpKind::GreaterEq, rhs)
+    }
+
+    pub fn less_equal(lhs: Expr<I>, rhs: Expr<I>) -> Self {
+        Self::cmp(lhs, CmpKind::LessEq, rhs)
+    }
+
+    pub fn cmp(lhs: Expr<I>, kind: CmpKind, rhs: Expr<I>) -> Self {
+        Self {
+            kind: ExprKind::Cmp(Box::new(lhs), kind, Box::new(rhs)),
+        }
+    }
+
+    pub fn and(lhs: Expr<I>, rhs: Expr<I>) -> Self {
+        Self {
+            kind: ExprKind::And(Box::new(lhs), Box::new(rhs)),
+        }
+    }
+
+    pub fn or(lhs: Expr<I>, rhs: Expr<I>) -> Self {
+        Self {
+            kind: ExprKind::Or(Box::new(lhs), Box::new(rhs)),
+        }
+    }
+
+    pub fn ret(ret: Option<Expr<I>>) -> Self {
+        Self {
+            kind: ExprKind::Return(ret.map(Box::new)),
+        }
+    }
+
+    pub fn brk(brk: Option<Expr<I>>) -> Self {
+        Self {
+            kind: ExprKind::Break(brk.map(Box::new)),
+        }
+    }
+
+    pub fn cont(cont: Option<Expr<I>>) -> Self {
+        Self {
+            kind: ExprKind::Continue(cont.map(Box::new)),
+        }
+    }
+
+    pub fn shr(lhs: Expr<I>, rhs: Expr<I>) -> Self {
+        Self::bin_op(lhs, BinaryOperand::Shr, rhs)
+    }
+
+    pub fn shl(lhs: Expr<I>, rhs: Expr<I>) -> Self {
+        Self::bin_op(lhs, BinaryOperand::Shl, rhs)
+    }
+
+    pub fn concat(lhs: Expr<I>, rhs: Expr<I>) -> Self {
+        Self::bin_op(lhs, BinaryOperand::Concat, rhs)
+    }
+
+    pub fn bit_or(lhs: Expr<I>, rhs: Expr<I>) -> Self {
+        Self::bin_op(lhs, BinaryOperand::BitOr, rhs)
+    }
+
+    pub fn bit_and(lhs: Expr<I>, rhs: Expr<I>) -> Self {
+        Self::bin_op(lhs, BinaryOperand::BitAnd, rhs)
+    }
+
+    pub fn rem(lhs: Expr<I>, rhs: Expr<I>) -> Self {
+        Self::bin_op(lhs, BinaryOperand::Rem, rhs)
+    }
+
+    pub fn if_(cond: Expr<I>, then: Expr<I>, else_: Expr<I>) -> Self {
+        Self {
+            kind: ExprKind::If(Box::new(cond), Box::new(then), Box::new(else_)),
+        }
+    }
+
+    pub fn bin_op(lhs: Expr<I>, op: BinaryOperand, rhs: Expr<I>) -> Self {
+        Self {
+            kind: ExprKind::BinOp(Box::new(lhs), op, Box::new(rhs)),
+        }
+    }
+
+    pub const fn is_nested(&self) -> bool {
+        matches!(self.kind, ExprKind::Nested(_))
+    }
+
+    pub const fn is_empty(&self) -> bool {
+        matches!(self.kind, ExprKind::Empty)
+    }
+
+    pub const fn is_literal(&self) -> bool {
+        matches!(self.kind, ExprKind::Literal(_))
+    }
+
+    pub fn simplify_nesting(self) -> Self {
+        NestingSimplifier::new().rewrite(self)
+    }
+
+    pub fn simplify_blocks(self) -> Self {
+        BlockSimplifier::new().rewrite(self)
+    }
+}
+
+impl<I> Default for Expr<I> {
+    fn default() -> Self {
+        Self {
+            kind: ExprKind::Empty,
+        }
+    }
+}
+
+#[derive(Clone, PartialEq, Eq, Hash)]
+pub enum ExprKind<I> {
+    Wildcard,
+    Empty,
```

Comment:
Where is `Empty` useful?

Reply:
It's used within the rewrites to remove/simplify expressions and it resolves to unit as its type